tests/test_packages/test_skills/test_generic_buyer/test_behaviours.py

```diff
@@ -0,0 +1,302 @@
+        )
+        assert has_attributes, error_str
+
+    def test_search_behaviour_act_not_is_searching(self):
+        """Test the act method of the search behaviour where is_searching is False."""
+        # setup
+        strategy = cast(GenericStrategy, self.skill.skill_context.strategy)
+        strategy._is_searching = False
+        search_behaviour = cast(
+            GenericSearchBehaviour, self.skill.skill_context.behaviours.search
+        )
+
+        # operation
+        search_behaviour.act()
+
+        # after
+        assert self.get_quantity_in_outbox() == 0
+
+
+class TestSkillHandler(BaseSkillTestCase):
+    """Test handlers of generic buyer."""
+
+    path_to_skill = Path(ROOT_DIR, "packages", "fetchai", "skills", "generic_buyer")
+
+    def test_fipa_handler_handle_unidentified_dialogue(self):
+        """Test the _handle_unidentified_dialogue method of the fipa handler."""
+        # setup
+        fipa_handler = cast(GenericFipaHandler, self.skill.skill_context.handlers.fipa)
+        incoming_message = self.build_incoming_message(
+            message_type=FipaMessage,
+            dialogue_reference=("", ""),
+            performative=FipaMessage.Performative.ACCEPT,
+        )
+
+        # operation
+        fipa_handler.handle(incoming_message)
+
+        # after
+        assert self.get_quantity_in_outbox() == 1, "No message in outbox."
+        has_attributes, error_str = self.message_has_attributes(
+            actual_message=self.get_message_from_outbox(),
+            message_type=DefaultMessage,
+            performative=DefaultMessage.Performative.ERROR,
+            to=incoming_message.sender,
+            sender=self.skill.skill_context.agent_address,
+            error_code=DefaultMessage.ErrorCode.INVALID_DIALOGUE,
+            error_msg="Invalid dialogue.",
+            error_data={"fipa_message": incoming_message.encode()},
+        )
+        assert has_attributes, error_str
+
+    def test_fipa_handler_handle_propose(self):
+        """Test the _handle_propose method of the fipa handler."""
+        # ToDo need to mock affordable and acceptable values
+        # setup
+        fipa_handler = cast(GenericFipaHandler, self.skill.skill_context.handlers.fipa)
+        fipa_dialogues = cast(FipaDialogues, self.skill.skill_context.fipa_dialogues)
+        proposal = Description(
+            {
+                "ledger_id": "some_ledger_id",
+                "price": "some_price",
+                "currency_id": "some_currency_id",
+                "service_id": "some_service_id",
+                "quantity": "some_quantity",
+                "tx_nonce": "some_tx_nonce",
+            }
+        )
+        incoming_message = self.prepare_dialogue(
+            dialogues=fipa_dialogues,
+            messages=((FipaMessage.Performative.CFP, {"query": "some_query"}),),
+            incoming_message=(FipaMessage.Performative.PROPOSE, {"proposal": proposal}),
+        )
```

Comment:
I did it slightly differently. Instead of the above, you first prepare the dialogue, then build the incoming_message based on it:

```
dialogue = self.prepare_dialogue(dialogues, messages=(...))
incoming_message = self.build_incoming_message(dialogue, ...)
```